### PR TITLE
Add Support for Database

### PR DESCRIPTION
**Pull request by foreverXZC**
Regarding to issue #136 #128 , I would like to introduce some functions related to database. Now only MySQL, PostgreSQL and Microsoft SQL Server are supported.

However, I just have tested these functions on Azure, not on AWS. Deploying an database on Azure is quite simple, but on AWS we have to use RDS, which requires more efforts. If needed, I can provide my test code for Azure. I think I need some time to deal with AWS cases.

### CODE REVIEW

File: modules/database/database.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+// DBQueryWithValidation queries from database and validate whether the result is the same as expected text. If there's any error, fail the test.
+func DBQueryWithValidation(t *testing.T, db *sql.DB, command string, expected string) {
+	err := DBQueryWithValidationE(t, db, command, expected)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// DBQueryWithValidationE queries from database and validate whether the result is the same as expected text. If not, return an error.
+func DBQueryWithValidationE(t *testing.T, db *sql.DB, command string, expected string) error {
```
[the remark this answers is not in the document]
Comment (foreverXZC) [reply modules/database/database.go]: Now `expected` here is a string. The function validates whether elements in one column are all the same as expected string. For instance, I insert two rows whose values are all ('abc'). Then I set `expected` as 'abc' and query from the table to check whether the result matches. 

However, we know that it is very specific and cannot be used in general. It should be a better choice to use `DBQueryWithCustomValidation`, where users can provide their own validation function.

File: modules/database/database.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"testing"
+
+	// Microsoft SQL Database Driver
+	_ "github.com/denisenkom/go-mssqldb"
+
+	// PostgreSQL Database Driver
+	_ "github.com/lib/pq"
+
+	// MySQL Database Driver
+	_ "github.com/go-sql-driver/mysql"
```
[the remark this answers is not in the document]
Comment (brikis98) [reply modules/database/database.go]: That sounds a bit odd. Typically, if you checkout a project and run `dep ensure`, all it should do is download the dependencies for that project into the `vendor` folder. You can then run `dep ensure -add github.com/go-sql-driver/mysql` to add a dependency, which will modify both `Gopkg` files.

[the remark this answers is not in the document]
Comment (foreverXZC) [reply modules/database/database.go]: My workflow is, first directing to database directory and then running `dep ensure`. After that, all dependencies including `gruntwork-io/terratest` are downloaded into vendor. `Gopkg.lock` becomes really big but 'Gopkg.toml' does not change. If I would like to execute `dep ensure -add`, it tells me that the dependency is already imported or required, so -add is only valid with a constraint, which means that adding dependencies fails. I wonder whether I'm doing something incorrectly.

[the remark this answers is not in the document]
Comment (mcalhoun) [reply modules/database/database.go]: Hi @foreverXZC, you should run `dep ensure` in the root of `terratest` (where the existing `Gopkg.toml` and `Gopkg.lock` files are). `dep` will find all the dependencies below that and will make sure they are installed for the project.

[the remark this answers is not in the document]
Comment (foreverXZC) [reply modules/database/database.go]: Hi @mcalhoun , after I run `dep ensure` in the root of terratest, `Gopkg.lock` becomes long and contains terratest packages. I think that's because test examples requires these things, so `dep ensure` install them into `Gopkg.lock`. My suggestion is that I only keep `modules` folder and then run `dep ensure`, so terratest packages will not be included in `Gopkg.lock`. What's your opinion?

**Comment by foreverXZC** [comment]
In addition, I have written test code to check whether the database runs correctly. As I have mentioned before, all Azure resources can be tested properly. For AWS, if I follow this link to deploy a MySQL Database, my test code also works.
https://aws.amazon.com/getting-started/tutorials/create-mysql-db/

However, I cannot find a proper method to provision a database on AWS using terraform. The MySQL Database created using following link seems do not work.
https://github.com/terraform-aws-modules/terraform-aws-rds/tree/v1.19.0/examples/complete-mysql
I think that's because it only creates a database, but is not responsible for making it accessible by users. For instance, I even cannot use MySQL Workbench or command line to connect to the database. Thus all test fails. Maybe security groups cause that problem.

As a result, I would like to know whether it is necessary for me to provide such terraform files that provision database on AWS which can be accessible by users. If yes, I think I really need help to achieve this.

File: modules/database/database.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcalhoun) [reply modules/database/database.go]: Hi @foreverXZC...I believe the reason you are seeing the terratest packages included in `Gopkg.lock` is because they are imported with fully qualified references to our repos ("github.com/gruntwork-io/terratest/...") and your repo is `foreverXZC/terratest` so dep sees those as external dependencies. This is normal behavior.

I will have to do a bit of research on the best way for us to merge your PR (any PR that includes dependency changes). I guess we will have to run `dep ensure` on the merged master branch (which will remove the external dependencies on `gruntwork-io/terratest`) and then commit the `Gopkg.lock` back to the master branch.

The tl;dr is run `dep ensure` on from the root of your repo, do work commit it to your PR and we'll deal with it when we merge (unless @brikis98 or someone else has a better idea of how to handle this).

**Comment by mcalhoun** [comment]
> As a result, I would like to know whether it is necessary for me to provide such terraform files that provision database on AWS which can be accessible by users. If yes, I think I really need help to achieve this.

@foreverXZC as @brikis98 mentioned above, the best way to test this would be to start a version of each database you want to test in docker as part of the test.

**Comment by foreverXZC** [comment]
Thanks for instructions. I know that if we use docker to create a database, it is easier for us to test it. However, using docker seems not so related to terraform and we also do not need to register for AWS, which is quite different from other existing examples. If it actually meets our requirement, do I need to put Dockerfile in examples folder? Thank you so much.

**Comment by ellisonc** [comment]
@foreverXZC, thanks for your contribution. We are removing old PRs. If this change is still required, we encourage you to re-open.
